recipes/full_finetune_distributed.py

```diff
@@ -657,12 +704,18 @@ def train(self) -> None:
                 # Step with optimizer
                 if (idx + 1) % self._gradient_accumulation_steps == 0:
                     if self._clip_grad_norm is not None:
-                        grad_norm = torch.nn.utils.clip_grad_norm_(
-                            self._model.parameters(),
-                            max_norm=float(self._clip_grad_norm),
-                        )
-                    self._optimizer.step()
-                    self._optimizer.zero_grad(set_to_none=True)
+                        if self._optimizer_in_bwd:
+                            raise NotImplementedError(
+                                "Gradient clipping is not supported after optimizer-in-the-backward."
+                            )
```

Comment:
Is this actually due to optimizer in backward, or something else? I don't think we have such a check in the single-device recipe

Reply:
It's due to optimizer-in-backward.
optimizer-in-backward calls `.step()` and `.zero_grad()` during `loss.backward()`, thus `grad`s are None at the time after to call `torch.nn.utils.clip_grad_norm_`, could not have gradient clipping successfully.
Single_device has the same issue that if `optimizer_in_bwd=True` and `clip_grad_norm=True`